### PR TITLE
Support in-place template reinstalls - for testing

### PR DESCRIPTION
**Pull request by ttasket**
This doesn't yet prevent appvms from starting with invalid template during the reinstall, and doesn't deal with the Netvm setting problem.
For issue #2061

### CODE REVIEW

**Comment by marmarek** [comment]
Probably both issues should be handled in `template.spec`:
- netvm - save previous value (and whether it was default or not)
- AppVM starting - it should be easy not need this at all - by extracting `root.img` to some temporary name and then atomically renaming over original `root.img`

**Comment by marmarek** [comment]
If you agree, I'd merge this PR and deal with those problems in the other one.

Hmm, on the other hand, changes to `template.spec` are slow to propagate - because templates are already in repository and contain %post script - no way to change those. And when user calls `reinstall` action, yum/dnf will try to get the same version, not newer. So maybe better to handle all of it here?

**Comment by ttasket** [comment]
We can deal with it here, perhaps easily... Was going to look for a lock() mechanism against `qvm-start`, but can simply invalidate the template during that time by renaming root.img (and maybe *.conf) to something else as the first step. What do you think?

**Comment by marmarek** [comment]
> but can simply invalidate the template during that time by renaming root.img to something else as the first step.

Yes, something like this. In fact, you already remove `root.img` in %post script of `template.spec` - basides the problem with old %post scripts in packages, it should be enough. So maybe indeed move it to `root.img.bak` here and if reinstallation succeed, remove it completely (and if not - restore it).

> (and maybe *.conf)

No need, it is autogenerated at every VM startup anyway.

**Comment by ttasket** [comment]
Right, this is for the sake of non-refreshed %post scriptlets. Doing `rm -f` here might be better than a rename because any interruption won't leave the old root image behind. However, that would place deletion before the yum "Yes/No" prompt. :(

Other thing is, doing it here I'm assuming I'll need to extrapolate between the package name and what paths are in the qubes db.

**Comment by marmarek** [comment]
> However, that would place deletion before the yum "Yes/No" prompt. :(

This is why I've proposed to rename it instead of removing. And also for cases like download fail etc. To not end up without any template...

> Other thing is, doing it here I'm assuming I'll need to extrapolate between the package name and what paths are in the qubes db.

`qvm-prefs fedora-23 root_img`

**Comment by ttasket** [comment]
OK.

Is it realistic the user will try to specify *version* in the package name? Can I use '.' or '.noarch' as delimiter?

**Comment by marmarek** [comment]
> Is it realistic the user will try to specify version in the package name? Can I use '.' or '.noarch' as delimiter?

As the template reinstall is very specific operation, I think you can expect to have exact template package name and nothing else. And fail for anything else matching `qubes-template-*` but you can't find existing template (like exact version, or wildcard).

Hmm, maybe better move this discussion back to qubesos/qubes-issues#2061 ?

**Comment by ttasket** [comment]
I agree with 'fail on fancy' here... we can treat reinstall of templates and dom0 components as separate things.

So I'll require just a single matching package in case of qubes-template-*.

File: dom0-updates/qubes-dom0-update [hunk 1]
```diff
@@ -173,8 +186,29 @@ elif [ -f /var/lib/qubes/updates/repodata/repomd.xml ]; then
...
+                    echo "OK"
+                else
+                    echo; echo "ERROR: Could not rename root.img"
+                    exit 1
+                fi
+            fi
+
+            yum $YUM_OPTS $YUM_ACTION ; RETCODE=$?
+            
+            if [[ -n "$BAK_TEMPLATE_ROOT" ]] ; then
```
Comment (marmarek) [inline dom0-updates/qubes-dom0-update]: Shouldn't this check be about `$TEMPLATE_NETVM`?

Comment (ttasket) [inline dom0-updates/qubes-dom0-update]: Intention is to repeat the check at ln191: It means, "are we handling a template?"

Comment (ttasket) [inline dom0-updates/qubes-dom0-update]: See new PR https://github.com/ttasket/qubes-core-admin-linux/pull/1

**Pull request merged**